manager/controllers/app/select_modules.go

```diff
@@ -31,199 +33,342 @@ type DataInfo struct {
+// TODO(shlomitk1): add plugins/transformation capabilities to this structure
+type ResolvedEdge struct {
+	Edge
+	Actions              []taxonomymodels.Action
+	Cluster              string
+	StorageAccountRegion string
+}
+
+// Solution is a final solution enabling a plotter construction.
+// It represents a full data flow between the data source and the workload.
+type Solution struct {
+	DataPath []ResolvedEdge
+}
+
+// FindPaths finds all valid data paths between the data source and the workload
+// First, data paths are constructed using interface connections, starting from data source.
+// Then, transformations are added to the found paths, and clusters are matched to satisfy restrictions from admin config policies.
+// Optimization is done by the shortest path (the paths are sorted by the length). To be changed in future versions.
+func (p *PlotterGenerator) FindPaths(item *DataInfo, appContext *app.FybrikApplication) []Solution {
+	// data source as appears in the asset metadata
+	source := Node{Connection: &item.DataDetails.Interface, Virtual: false}
+	// data sink, either a virtual endpoint in read scenarios, or a datastore as in ingest scenario
+	destination := Node{Connection: &item.Context.Requirements.Interface, Virtual: (appContext.Spec.Selector.WorkloadSelector.Size() > 0)}
+	// find data paths of length up to 2 from data source to the workload, not including transformations or branches
+	solutions := p.findPathsWithinLimit(item, &source, &destination, 2)
+	// get valid solutions by extending data paths with transformations and selecting an appropriate cluster for each capability
+	solutions = p.validSolutions(item, solutions, appContext)
+	return solutions
+}
+
+// extend the received data paths with transformations and select an appropriate cluster for each capability in a data path
+func (p *PlotterGenerator) validSolutions(item *DataInfo, solutions []Solution, appContext *app.FybrikApplication) []Solution {
+	validPaths := []Solution{}
+	for ind := range solutions {
+		if p.validate(item, solutions[ind], appContext) {
+			validPaths = append(validPaths, solutions[ind])
+		}
+	}
+	return validPaths
+}
+
+func (p *PlotterGenerator) validate(item *DataInfo, solution Solution, appContext *app.FybrikApplication) bool {
+	// start from data source, check supported actions and cluster restrictions
+	requiredActions := item.Actions
+	for ind := range solution.DataPath {
+		element := &solution.DataPath[ind]
+		element.Actions = []taxonomymodels.Action{}
+		moduleCapability := element.Module.Spec.Capabilities[element.CapabilityIndex]
+		if !element.Edge.Sink.Virtual {
+			// storage is required, plus more actions on copy may be needed
+			for _, region := range p.StorageAccountRegions {
+				// query the policy manager whether WRITE operation is allowed
+				operation := new(taxonomymodels.PolicyManagerRequestAction)
+				operation.SetActionType(taxonomymodels.WRITE)
```

Comment:
We also wrote copy policies.  How do you decide whether to check for write and not copy?

Reply:
There are no copy policies.
https://github.com/fybrik/fybrik/blob/2af0fe8f0a82ef80d6133c705ab0112f89967fe6/pkg/taxonomy/model/policymanager/base/model_action_type.go#L21-L26

Reply:
charts/fybrik/files/adminconfig/default_policies.rego has a copy policy.  I guess you are differentiating between actions (read, write and delete) and copy as a capability.  Correct?

Reply:
Correct.